### PR TITLE
Implement C++20 `<source_location>`

### PR DESCRIPTION
**Pull request by miscco**
This implements C++20 `source_location` feature

We had some issues with source_location coming in through CUDASTF, so put it into libcu++ proper

### CODE REVIEW

**Comment by miscco** [comment]
@alliepiper Somehow cudastf is not picking up the libcu++ sources, can you give that a look, I thought I did the right thing

File: libcudacxx/include/cuda/std/__cccl/builtin.h [hunk 1]
```diff
@@ -135,6 +147,32 @@
...
+#  define _CCCL_BUILTIN_FILE() __builtin_FILE()
+#else // ^^^ _CCCL_HAS_BUILTIN(__builtin_FILE) ^^^ / vvv !_CCCL_HAS_BUILTIN(__builtin_FILE) vvv
+#  define _CCCL_BUILTIN_FILE() __FILE__
+#endif // !_CCCL_HAS_BUILTIN(__builtin_LINE)
+
+// NVCC below 11.3 cannot handle __builtin_FILE
+#if defined(_CCCL_CUDACC_BELOW_11_3)
+#  undef _CCCL_BUILTIN_FILE
+#  define _CCCL_BUILTIN_FILE() __FILE__
+#endif // _CCCL_CUDACC_BELOW_11_3
```
Comment (andralex) [inline libcudacxx/include/cuda/std/__cccl/builtin.h:161]: Similarly with above, one condition should cover the two possibilities.

File: libcudacxx/include/cuda/std/__cccl/builtin.h [hunk 2]
```diff
@@ -135,6 +147,32 @@
...
+#  define _CCCL_BUILTIN_FUNCTION() __builtin_FUNCTION()
+#else // ^^^ _CCCL_HAS_BUILTIN(__builtin_FUNCTION) ^^^ / vvv !_CCCL_HAS_BUILTIN(__builtin_FUNCTION) vvv
+#  define _CCCL_BUILTIN_FUNCTION() "__builtin_FUNCTION is unsupported!"
+#endif // !_CCCL_HAS_BUILTIN(__builtin_FUNCTION)
+
+// NVCC below 11.3 cannot handle __builtin_FUNCTION
+#if defined(_CCCL_CUDACC_BELOW_11_3)
+#  undef _CCCL_BUILTIN_FUNCTION
+#  define _CCCL_BUILTIN_FUNCTION() "__builtin_FUNCTION is unsupported!"
+#endif // _CCCL_CUDACC_BELOW_11_3
```
Comment (andralex) [inline libcudacxx/include/cuda/std/__cccl/builtin.h:174]: here too

File: libcudacxx/include/cuda/std/__cccl/builtin.h [hunk 3]
```diff
@@ -156,6 +194,19 @@
...
+#  define _CCCL_BUILTIN_LINE() __builtin_LINE()
+#else // ^^^ _CCCL_HAS_BUILTIN(__builtin_LINE) ^^^ / vvv !_CCCL_HAS_BUILTIN(__builtin_LINE) vvv
+#  define _CCCL_BUILTIN_LINE() __LINE__
+#endif // !_CCCL_HAS_BUILTIN(__builtin_LINE)
+
+// NVCC below 11.3 cannot handle __builtin_LINE
+#if defined(_CCCL_CUDACC_BELOW_11_3)
+#  undef _CCCL_BUILTIN_LINE
+#  define _CCCL_BUILTIN_LINE() __LINE__
+#endif // _CCCL_CUDACC_BELOW_11_3
```
Comment (andralex) [inline libcudacxx/include/cuda/std/__cccl/builtin.h:208]: and last one here

File: libcudacxx/include/cuda/std/__cccl/builtin.h [hunk 4]
```diff
@@ -135,6 +147,32 @@
...
+#if defined(_CCCL_CUDACC_BELOW_11_3)
+#  undef _CCCL_BUILTIN_FILE
+#  define _CCCL_BUILTIN_FILE() __FILE__
+#endif // _CCCL_CUDACC_BELOW_11_3
+
+#if _CCCL_HAS_BUILTIN(__builtin_FUNCTION) || defined(_CCCL_COMPILER_GCC) \
+  || (defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION >= 1927)
+#  define _CCCL_BUILTIN_FUNCTION() __builtin_FUNCTION()
+#else // ^^^ _CCCL_HAS_BUILTIN(__builtin_FUNCTION) ^^^ / vvv !_CCCL_HAS_BUILTIN(__builtin_FUNCTION) vvv
+#  define _CCCL_BUILTIN_FUNCTION() "__builtin_FUNCTION is unsupported!"
```
Comment (andralex) [inline libcudacxx/include/cuda/std/__cccl/builtin.h:167]:
```suggestion
#  define _CCCL_BUILTIN_FUNCTION() "__builtin_FUNCTION is unsupported"
```

File: libcudacxx/include/cuda/std/source_location [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+//! NOTE: we cannot use `__builtin_source_location` as that requires us to use `::std::source_location::impl`
+class source_location
+{
+private:
+  const char* __file_name_     = "";
+  const char* __function_name_ = "";
+  uint_least32_t __line_       = 0;
+  uint_least32_t __column_     = 0;
```
Comment (andralex) [inline libcudacxx/include/cuda/std/source_location:36]: If there's any way to get away with it, any chance we stick with conventional names for these members (no double underscore)?

File: libcudacxx/include/cuda/std/__cccl/builtin.h [hunk 6]
```diff
@@ -127,6 +127,18 @@
...
+#if _CCCL_HAS_BUILTIN(__builtin_COLUMN) || defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION >= 1927
+#  define _CCCL_BUILTIN_COLUMN() __builtin_COLUMN()
+#else // ^^^ _CCCL_HAS_BUILTIN(__builtin_COLUMN) ^^^ / vvv !_CCCL_HAS_BUILTIN(__builtin_COLUMN) vvv
+#  define _CCCL_BUILTIN_COLUMN() 0
+#endif // !_CCCL_HAS_BUILTIN(__builtin_COLUMN)
+
+// NVCC below 11.3 cannot handle __builtin_COLUMN
+#if defined(_CCCL_CUDACC_BELOW_11_3)
+#  undef __builtin_COLUMN
+#  define __builtin_COLUMN() 0
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__cccl/builtin.h:139]:
```suggestion
#  undef _CCCL_BUILTIN_COLUMN
#  define _CCCL_BUILTIN_COLUMN() 0
```

File: libcudacxx/include/cuda/std/__cccl/builtin.h [hunk 7]
```diff
@@ -127,6 +127,18 @@
 #  undef _CCCL_BUILTIN_BIT_CAST
 #endif // clang < 10 || nvcc < 11.7
 
+#if _CCCL_HAS_BUILTIN(__builtin_COLUMN) || defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION >= 1927
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__cccl/builtin.h:130]:
```suggestion
#if _CCCL_HAS_BUILTIN(__builtin_COLUMN) || (defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION >= 1927)
```

File: libcudacxx/include/cuda/std/version [hunk 8]
```diff
@@ -54,7 +54,8 @@
 #  ifndef _LIBCUDACXX_HAS_NO_THREADS
 // #   define __cccl_lib_shared_timed_mutex                 201402L
 #  endif // !_LIBCUDACXX_HAS_NO_THREADS
-#  define __cccl_lib_span 202002L
+#  define __cccl_lib_source_location 201907L
+#  define __cccl_lib_span            202002L
```
Comment (andralex) [inline libcudacxx/include/cuda/std/version:58]: careful, this seems unrelated

Comment (andralex) [reply libcudacxx/include/cuda/std/version:58]: nvm it's just the alignment

File: libcudacxx/include/cuda/std/__cccl/builtin.h [hunk 9]
```diff
@@ -127,6 +127,18 @@
...
+#  define _CCCL_BUILTIN_COLUMN() __builtin_COLUMN()
+#else // ^^^ _CCCL_HAS_BUILTIN(__builtin_COLUMN) ^^^ / vvv !_CCCL_HAS_BUILTIN(__builtin_COLUMN) vvv
+#  define _CCCL_BUILTIN_COLUMN() 0
+#endif // !_CCCL_HAS_BUILTIN(__builtin_COLUMN)
+
+// NVCC below 11.3 cannot handle __builtin_COLUMN
+#if defined(_CCCL_CUDACC_BELOW_11_3)
+#  undef __builtin_COLUMN
+#  define __builtin_COLUMN() 0
+#endif // _CCCL_CUDACC_BELOW_11_3
```
[the remark this answers is not in the document]
Comment (miscco) [reply libcudacxx/include/cuda/std/__cccl/builtin.h:140]: This is a convention we started because it becomes really difficult to track what is enabled and what is not at times.

So we settled on a convention that we have one condition that enables something and one condition that explicitly disables something if there is a compiler bug

File: libcudacxx/include/cuda/std/__cccl/builtin.h (hunk 1 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/__cccl/builtin.h:161]: Ditto that makes things easier

File: libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+static_assert(empty.column() == 0, "");
+static_assert(empty.file_name()[0] == '\0', "");
+static_assert(empty.function_name()[0] == '\0', "");
+
+ASSERT_NOEXCEPT(empty.line());
+ASSERT_NOEXCEPT(empty.column());
+ASSERT_NOEXCEPT(empty.file_name());
+ASSERT_NOEXCEPT(empty.function_name());
+static_assert(cuda::std::is_same<unsigned, decltype(empty.line())>::value, "");
+static_assert(cuda::std::is_same<unsigned, decltype(empty.column())>::value, "");
```
Comment (andralex) [inline libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp:49]:
```suggestion
static_assert(cuda::std::is_same<::std:::uint_least32_t, decltype(empty.line())>::value, "");
static_assert(cuda::std::is_same<::std:::uint_least32_t, decltype(empty.column())>::value, "");
```

File: libcudacxx/include/cuda/std/source_location (hunk 5 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/source_location:36]: Done

File: libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+static_assert(cuda::std::is_same<const char*, decltype(empty.file_name())>::value, "");
+static_assert(cuda::std::is_same<const char*, decltype(empty.function_name())>::value, "");
+
+__device__ _CCCL_CONSTEXPR_GLOBAL cuda::std::source_location device_empty{};
+#if !defined(_CCCL_CUDACC_BELOW_11_3)
```
Comment (andralex) [inline libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp:54]: Wouldn't the checks below also hold for cudacc below 11.3?

File: libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+ASSERT_NOEXCEPT(device_empty.line());
+ASSERT_NOEXCEPT(device_empty.column());
+ASSERT_NOEXCEPT(device_empty.file_name());
+ASSERT_NOEXCEPT(device_empty.function_name());
+
+// A simple use of current() outside a function.
+#line 1000 "ss"
+constexpr cuda::std::source_location cur = cuda::std::source_location::current();
+static_assert(cur.line() == 1000, "");
```
Comment (andralex) [inline libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp:69]: This will mess up file and line for the rest of the file, leading to confusing location for all error messages below. Unfortunately saving `__FILE__` and `__LINE__` is not possible, or at least I couldn't find a way. I tried this:

```cpp
#define SAVE_LINE __LINE__
#line 1000
#line SAVE_LINE
```

So I suggest this:

```cpp
constexpr my_namespace::source_location cur;
static_assert(cur.line() + 1 == __LINE__, "");
```

This will check exactly what you need, wherever the line number is.

File: libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+#if _CCCL_HAS_BUILTIN(__builtin_COLUMN) || defined(TEST_COMPILER_MSVC) && _CCCL_MSVC_VERSION >= 1927
+static_assert(cur.column() > 0, "");
+#else // ^^^ _CCCL_BULTIN_COLUMN ^^^ / vvv !_CCCL_BULTIN_COLUMN vvv
+static_assert(cur.column() == 0, "");
+#endif // !_CCCL_BULTIN_COLUMN
+static_assert(cur.file_name()[0] == 's' && cur.file_name()[1] == 's' && cur.file_name()[2] == '\0', "");
```
Comment (andralex) [inline libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp:76]:
```suggestion
static_assert(cur.file_name()[0] == __FILE__[0] && cur.file_name()[1] == __FILE__[1] && cur.file_name()[2] == __FILE__[sizeof(__FILE) - 1], "");
```

File: libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    if (target[index] == '\0')
+    {
+      return false;
+    }
+  }
+}
+
+__host__ __device__ void test()
+{
+#line 2000
```
Comment (andralex) [inline libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp:135]: This will also ruin line numbers below. Suggestion: forward declare a function such as:

```cpp
cuda::std::source_location fake_source_location();
```

Then call that function inside the unttest. At the VERY END of this file add this:

```cpp
#line 1000 "sss"
cuda::std::source_location fake_source_location() { return cuda::std::source_location::current(); }
// End of file
```

Then you have deterministic values for line and file.

File: libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general_nvrtc.pass.cpp [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    {
+      return false;
+    }
+
+    if (lhs[index] == '\0')
+    {
+      return true;
+    }
+  }
+}
```
Comment (andralex) [inline libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general_nvrtc.pass.cpp:44]: Is this code generated? Otherwise we'd do good to factor it in somewhere.

**Comment by andralex** [comment]
lgtm modulo comments

File: libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp (hunk 11 above)
Comment (miscco) [reply libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp:54]: This is specific towards `device_empty` not being a constant expression so we cannot use it within a static_assert

File: libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp (hunk 12 above)
Comment (miscco) [reply libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp:69]: That is correct, but we actually want to check that this is doing the right thing when someone uses that pragma

The test is really short and specific so I would like to keep it that way

File: libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp (hunk 14 above)
Comment (miscco) [reply libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp:135]: Ditto, it is testing intended functionality and if somethign breaks the test is short enough to work with it

File: libcudacxx/include/cuda/std/__cccl/builtin.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (andralex) [reply libcudacxx/include/cuda/std/__cccl/builtin.h:140]: Makes sense, thanks.

File: libcudacxx/include/cuda/std/__cccl/builtin.h (hunk 6 above)
Comment (ericniebler) [inline libcudacxx/include/cuda/std/__cccl/builtin.h:139]: this issue is still not fixed

File: libcudacxx/include/cuda/std/__cccl/builtin.h [hunk 16]
```diff
@@ -135,6 +147,32 @@
...
+#if defined(_CCCL_CUDACC_BELOW_11_3)
+#  undef _CCCL_BUILTIN_FILE
+#  define _CCCL_BUILTIN_FILE() __FILE__
+#endif // _CCCL_CUDACC_BELOW_11_3
+
+#if _CCCL_HAS_BUILTIN(__builtin_FUNCTION) || defined(_CCCL_COMPILER_GCC) \
+  || (defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION >= 1927)
+#  define _CCCL_BUILTIN_FUNCTION() __builtin_FUNCTION()
+#else // ^^^ _CCCL_HAS_BUILTIN(__builtin_FUNCTION) ^^^ / vvv !_CCCL_HAS_BUILTIN(__builtin_FUNCTION) vvv
+#  define _CCCL_BUILTIN_FUNCTION() "__builtin_FUNCTION is unsupported"
```
Comment (ericniebler) [inline libcudacxx/include/cuda/std/__cccl/builtin.h:167]: C99 and C++11 added `__func__`, all compilers i know have `__FUNCTION__`.

**Review by ericniebler — CHANGES_REQUESTED** [review]

File: libcudacxx/include/cuda/std/__cccl/builtin.h (hunk 16 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/__cccl/builtin.h:167]: The issue is that both macros are only allowed within the body of a function, see https://godbolt.org/z/s3Wcv5T11

File: libcudacxx/include/cuda/std/__cccl/builtin.h (hunk 6 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/__cccl/builtin.h:139]: What issue?

File: libcudacxx/include/cuda/std/__cccl/builtin.h (hunk 16 above)
Comment (ericniebler) [reply libcudacxx/include/cuda/std/__cccl/builtin.h:167]: what does `__builtin_FUNCTION()` do outside of a function? when would you want to use it like that?

File: libcudacxx/include/cuda/std/source_location [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX_SOURCE_LOCATION
+#define _LIBCUDACXX_SOURCE_LOCATION
```
Comment (ericniebler) [inline libcudacxx/include/cuda/std/source_location:12]: mild preference for giving our headers proper filename extensions.

File: libcudacxx/include/cuda/std/__cccl/builtin.h (hunk 16 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/__cccl/builtin.h:167]: Its doing what is needed for source location. All those macros were effectively used for that

File: cudax/cmake/cudaxBuildCompilerTargets.cmake [hunk 18]
```diff
@@ -11,6 +11,12 @@
 #   be linked into the developer build targets, as they include both
 #   cudax.compiler_interface and cccl.compiler_interface_cppXX.
 
+find_package(libcudacxx ${cudax_VERSION} EXACT CONFIG
```
Comment (alliepiper) [inline cudax/cmake/cudaxBuildCompilerTargets.cmake:14]:
```suggestion
find_package(libcudacxx ${cudax_VERSION} EXACT CONFIG REQUIRED
```

File: cudax/cmake/cudaxBuildCompilerTargets.cmake [hunk 19]
```diff
@@ -63,6 +69,7 @@ function(cudax_build_compiler_targets)
       # order matters here, we need the cudax options to override the cccl options.
       cccl.compiler_interface_cpp${dialect}
       cudax.compiler_interface
+      cudax.libcudacxx
```
Comment (alliepiper) [inline cudax/cmake/cudaxBuildCompilerTargets.cmake:72]:
```suggestion
      libcudacxx::libcudacxx
```

File: cudax/cmake/cudaxBuildCompilerTargets.cmake [hunk 20]
```diff
@@ -11,6 +11,12 @@
 #   be linked into the developer build targets, as they include both
 #   cudax.compiler_interface and cccl.compiler_interface_cppXX.
 
+find_package(libcudacxx ${cudax_VERSION} EXACT CONFIG
+  NO_DEFAULT_PATH # Only check the explicit path in HINTS:
+  HINTS "${CCCL_SOURCE_DIR}/lib/cmake/libcudacxx/"
+)
+thrust_create_target(cudax.libcudacxx)
```
Comment (alliepiper) [inline cudax/cmake/cudaxBuildCompilerTargets.cmake:18]: Don't use `thrust_create_target` here -- just use the `libcudacxx::libcudacxx` that is found by `find_package(libcudacxx ...)`.

```suggestion
```

**Review by alliepiper — DISMISSED** [review]

File: libcudacxx/include/cuda/std/source_location (hunk 17 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/source_location:12]: That ship has sailed with standard headers

**Review by ericniebler — APPROVED** [review]

File: libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general_nvrtc.pass.cpp (hunk 15 above)
Comment (miscco) [reply libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general_nvrtc.pass.cpp:44]: We have a overhaul of `char_traits` on our plate, so I am postponing that until then
